sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -151,6 +151,238 @@ object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
+          Literal.FalseLiteral
+        case c EqualNullSafe d
+          if planEqual(a, d) && (planEqual(b, c) || valueLessThanOrEqual(b, c)) =>
+          Literal.FalseLiteral
+
+        case c EqualNullSafe d if planEqual(b, d) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(b, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, d) => EqualTo(c, d)
+      }
+      case a LessThanOrEqual b => expression transformUp {
+        case c LessThan d if planEqual(b, d) && valueLessThan(c, a) =>
+          Literal.TrueLiteral
+        case c LessThan d if planEqual(b, c) && (planEqual(a, d) || valueLessThanOrEqual(d, a)) =>
+          Literal.FalseLiteral
+        case c LessThan d if planEqual(a, c) && valueLessThan(b, d) =>
+          Literal.TrueLiteral
+        case c LessThan d if planEqual(a, d) && (planEqual(b, c) || valueLessThanOrEqual(b, c)) =>
+          Literal.FalseLiteral
+
+        case c LessThanOrEqual d
+          if planEqual(b, d) && (planEqual(a, c) || valueLessThanOrEqual(c, a)) =>
+          Literal.TrueLiteral
+        case c LessThanOrEqual d if planEqual(b, c) && valueLessThan(d, a) =>
+          Literal.FalseLiteral
+        case c LessThanOrEqual d if planEqual(b, c) && (planEqual(a, d) || valueEqual(a, d)) =>
+          EqualTo(c, d)
+        case c LessThanOrEqual d
+          if planEqual(a, c) && (planEqual(b, d) || valueLessThanOrEqual(b, d)) =>
+          Literal.TrueLiteral
+        case c LessThanOrEqual d if planEqual(a, d) && valueLessThan(b, c) =>
+          Literal.FalseLiteral
+        case c LessThanOrEqual d if planEqual(a, d) && (planEqual(b, c) || valueEqual(b, c)) =>
+          EqualTo(c, d)
+
+        case c EqualTo d if planEqual(b, d) && valueLessThan(c, a) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(b, c) && valueLessThan(d, a) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(a, c) && valueLessThan(b, d) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(a, d) && valueLessThan(b, c) => Literal.FalseLiteral
+
+        case c EqualNullSafe d if planEqual(b, d) && valueLessThan(c, a) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(b, c) && valueLessThan(d, a) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(a, c) && valueLessThan(b, d) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(a, d) && valueLessThan(b, c) => Literal.FalseLiteral
+
+        case c EqualNullSafe d if planEqual(b, d) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(b, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, d) => EqualTo(c, d)
+      }
+      case a EqualTo b => expression transformUp {
+        case c LessThan d if planEqual(b, d) && planEqual(a, c) => Literal.FalseLiteral
```

Comment:
actually these too are wrong with `null`s, as they should return `null` instead of `false` and the difference is significant (consider the case of a `Not` in front of it..)

Reply:
Thanks! I think you are right and this is an issue at many other places too.

Reply:
I think the solution could be checking if `a`, `b`, `c` and `d` not nullable and then we can use `transform`. Otherwise limiting a top-down way replacement until we have any of `And`, `Or`, `Case` or `If` nodes and can't go deeper if we hit something else. This is what `ReplaceNullWithFalseInPredicate` rule does.

Reply:
Fixed the issue and added some new UTs.